v4/opt/factory.go

```diff
@@ -88,7 +107,7 @@ func (f *Factory) concatFilterConditions(filterLeft, filterRight GroupID) GroupI
 	return f.ConstructFilters(f.StoreList(items))
 }
 
-func (f *Factory) flattenFilterCondition(filter GroupID) GroupID {
+func (f *Factory) flattenFilterCondition(input ListID, filter GroupID) GroupID {
```

Comment:
Adding the check here might miss some cases. The trouble is that `flattenFilterCondition` is called once in order to turn a possibly nested select/join predicate into a flattened list. But what happens if other patterns add items to that list later on (e.g. pushdown patterns)? It wouldn't go through this code path, and so we wouldn't infer equivalent filters for the newly added expressions.

I think this is fine for opttoy, but maybe add a comment about the issue, to be solved later? We might decide to add separate patterns for inferring equivalent filters (for both select and join cases). I'm not quite sure how we can efficiently check whether we've already pushed down filters, because if we're not careful, we could repeatedly try to push down the same filters. Your solution avoids this.